src/pubsub.rs

```diff
@@ -0,0 +1,254 @@
+use std::collections::{HashMap, HashSet};
+
+use abscissa_core::{
+    tracing::log::{debug, error},
+    Application,
+};
+use somm_proto::{
+    cosmos_sdk_proto::cosmos::base::abci::v1beta1::TxResponse,
+    pubsub::{
+        query_client::QueryClient, QueryDefaultSubscriptionsRequest,
+        QueryDefaultSubscriptionsResponse, QueryPublisherRequest, QueryPublisherResponse,
+        QuerySubscriberIntentsBySubscriberAddressRequest,
+        QuerySubscriberIntentsBySubscriberAddressResponse, Subscriber,
+    },
+};
+use url::Url;
+use x509_parser::prelude::parse_x509_pem;
+
+use crate::{
+    config::get_delegate_address,
+    error::{Error, ErrorKind},
+    prelude::APP,
+    somm_send,
+};
+
+use self::cache::PublisherTrustData;
+
+pub(crate) mod cache;
+
+/// Retrieves the PEM encoded CA certs for all default subscription publishers.
+pub(crate) async fn get_trust_state() -> Result<Vec<PublisherTrustData<'static>>, Error> {
+    debug!("collecting trust roots from default subscriptions and subscriber intents.");
+    let config = APP.config();
+
+    // load subscription mappings
+    let mut subscription_mappings: HashMap<String, HashSet<String>> = HashMap::default();
+    get_default_subscriptions()
+        .await?
+        .default_subscriptions
+        .iter()
+        .for_each(|s| {
+            if config
+                .pubsub
+                .publisher_domain_block_list
+                .contains(&s.publisher_domain)
+            {
+                debug!(
+                    "publisher domain {} is in the block list. skipping.",
+                    s.publisher_domain
+                );
+
+                return;
+            }
+
+            subscription_mappings
+                // if there is an entry for the key, push to the vec and dedup,
+                // otherwise create a new vec with the subscription id
+                .entry(s.publisher_domain.to_owned())
+                .and_modify(|e| {
+                    e.insert(s.subscription_id.to_owned());
+                })
+                .or_insert(HashSet::from([s.subscription_id.to_owned()]));
+        });
+
+    let delegate_address = get_delegate_address().to_string();
+    get_subscriber_intents_by_subscriber_address(delegate_address)
+        .await?
+        .subscriber_intents
+        .iter()
+        .for_each(|si| {
+            if config
+                .pubsub
+                .publisher_domain_block_list
+                .contains(&si.publisher_domain)
+            {
+                debug!(
+                    "publisher domain {} is in the block list. skipping.",
+                    si.publisher_domain
+                );
+
+                return;
+            }
+
+            // if there is already a default entry for this subscription ID under
+            // a different publisher, we remove it from that set of IDs.
+            if let Some(p) = subscription_mappings.keys().find(|k| {
```

Comment:
I think what we're doing here is maybe the reverse order of what we want. We're creating subscription mappings with the publisher as the key, which means we have to do a bunch of weird searching. What if instead the subscription ID is the key, and the publisher is the value? Then it would look like this:

1. Write out default subscription mapping
2. Run through subscriberIntent mapping -- simply overwrite the publisher key, if the subscriber is using the default it won't change, if they've customized it it will

Reply:
Yeah that would be simpler